### PR TITLE
Genre id and fix

### LINKED ISSUE #109: Used space instead of free space in GetDriveSpace
If you you have a 1 TB hdd and 200 GB space are used, the client shows the text "200 GB of 1000 GB are available". I've checked it with Leia Beta 3 and 2 skins, both show the wrong text, so it should not be a skin issue.

I, for one, would not sum up the space if more than one hdd or stick is found, I would use only the biggest disk. The receiver also uses only one hdd for recordings, most of the time the other stick is only a swap-space or left from the last movie-night. Therefore the client shows more free space than effectively available. Not a problem if you have a lot of free space, but if you record way too much...

Thanks

### PR DESCRIPTION
**Pull request by phunkyfish**
v3.12.4
- Fixed: Used space instead of free space in GetDriveSpace - fixes #109
- Added: Genre id support from OTA feeds - Requires OpenWebIf 1.3.5

### CODE REVIEW

**Comment by phunkyfish** [comment]
I guarantee this one is easier to review than the last ;)

File: src/Enigma2.cpp [hunk 1]
```diff
@@ -1387,6 +1400,19 @@ PVR_ERROR Enigma2::GetEPGForChannel(ADDON_HANDLE handle, const PVR_CHANNEL &chan
       entry.SetPlotOutline("");
     }    
 
+    if (XMLUtils::GetString(pNode, "e2eventgenre", strTmp))
```
Comment (Jalle19) [inline src/Enigma2.cpp]: I'd move this somewhere so you wouldn't have to duplicate the code

Comment (phunkyfish) [reply src/Enigma2.cpp]: In fact that whole XML reading block is pretty much duplicated. Which is the focus of my next PR ;)

So will leave it as for the moment if you don't mind?

Comment (Jalle19) [reply src/Enigma2.cpp]: Sure

File: src/Enigma2.cpp [hunk 2]
```diff
@@ -1774,51 +1800,41 @@ PVR_ERROR Enigma2::GetDriveSpace(long long *iTotal, long long *iUsed)
...
+  {
+    std::regex regexSize ("^.* " + size);
+    std::regex regexReplaceSize (" " + size);
+
+    if (regex_match(stringInMbGbTb, regexSize))
     {
-      if (regex_match(freeSpace, regexMB))
-      {
-        double megabytesSize = atof(regex_replace(freeSpace, regexReplaceMB, replaceWith).c_str());
+      double sizeValue = atof(regex_replace(stringInMbGbTb, regexReplaceSize, replaceWith).c_str());
```
Comment (ksooo) [inline src/Enigma2.cpp]: Remove `replaceWith`variable and just use `""` here? string construction is not that performance critical here I guess.

File: src/Enigma2.cpp [hunk 3]
```diff
@@ -1774,51 +1800,41 @@ PVR_ERROR Enigma2::GetDriveSpace(long long *iTotal, long long *iUsed)
...
-    {
-      double gigabytesSize = atof(regex_replace(freeSpace, regexReplaceGB, replaceWith).c_str());
+long long Enigma2::GetKbFromString(const std::string &stringInMbGbTb) const
+{
+  long long sizeInKb = 0;
 
-      used += (long long)(gigabytesSize * 1024 * 1024);
-    }
-    else
+  std::vector<std::string> sizes = {"MB", "GB", "TB"};
```
Comment (ksooo) [inline src/Enigma2.cpp]: This vector could be `static const`. Then, it gets constructed only once.

File: src/Enigma2.cpp [hunk 4]
```diff
@@ -1774,51 +1800,41 @@ PVR_ERROR Enigma2::GetDriveSpace(long long *iTotal, long long *iUsed)
...
 
-      total += (long long)(gigabytesSize * 1024 * 1024);
-    }
-    else
-    {
-      //It's MB
-      if (regex_match(capacity, regexMB))
-      {
-        double megabytesSize = atof(regex_replace(capacity, regexReplaceMB, replaceWith).c_str());
+  Logger::Log(LEVEL_INFO, "GetDriveSpace Total: %u, Used %u", *iTotal, *iUsed);
```
Comment (ksooo) [inline src/Enigma2.cpp]: Are you sure printf'ing a `long long`usinf `%u` works? Isn't it `%lld` or that PRId64 stuff from inttypes.h ?

File: src/Enigma2.cpp [hunk 5]
```diff
@@ -1774,51 +1800,41 @@ PVR_ERROR Enigma2::GetDriveSpace(long long *iTotal, long long *iUsed)
...
+{
+  long long sizeInKb = 0;
 
-      used += (long long)(gigabytesSize * 1024 * 1024);
-    }
-    else
+  std::vector<std::string> sizes = {"MB", "GB", "TB"};
+  long multiplier = 1024 * 1024;
+  std::string replaceWith = "";
+  for (auto& size : sizes)
```
Comment (ksooo) [inline src/Enigma2.cpp]: As `size` is not a complex type, but just `std::string`, II would not use `auto` here. But ofc this is just personal taste.

Bit it should read `const --choose-your-type-&`. ;-)

File: src/Enigma2.cpp [hunk 6]
```diff
@@ -1774,51 +1800,41 @@ PVR_ERROR Enigma2::GetDriveSpace(long long *iTotal, long long *iUsed)
...
-      if (regex_match(freeSpace, regexMB))
-      {
-        double megabytesSize = atof(regex_replace(freeSpace, regexReplaceMB, replaceWith).c_str());
+      double sizeValue = atof(regex_replace(stringInMbGbTb, regexReplaceSize, replaceWith).c_str());
 
-        used += (long long)(megabytesSize * 1024);
-      }
-    }
-  }
+      sizeInKb += (long long)(sizeValue * multiplier);
```
[the remark this answers is not in the document]
Comment (phunkyfish) [reply src/Enigma2.cpp]: Done

File: src/Enigma2.cpp (hunk 2 above)
Comment (phunkyfish) [reply src/Enigma2.cpp]: Done

File: src/Enigma2.cpp (hunk 3 above)
Comment (phunkyfish) [reply src/Enigma2.cpp]: I should have spotted that.

File: src/Enigma2.cpp (hunk 5 above)
Comment (phunkyfish) [reply src/Enigma2.cpp]: Done

File: src/Enigma2.cpp (hunk 4 above)
Comment (phunkyfish) [reply src/Enigma2.cpp]: Done

File: src/Enigma2.cpp (hunk 2 above)
Comment (phunkyfish) [reply src/Enigma2.cpp]: This actually failed in Travis but worked locally. Had to revert it.

File: src/enigma2/extract/GenreExtractor.cpp [hunk 7]
```diff
@@ -25,24 +25,27 @@ GenreExtractor::~GenreExtractor(void)
...
-  if (!genreText.empty() && genreText != GENRE_RESERVED_IGNORE)
+  if (entry.GetGenreType() == 0)
   {
-    int combinedGenreType = GetGenreTypeFromText(genreText, entry.GetTitle());
-
-    if (combinedGenreType == EPG_EVENT_CONTENTMASK_UNDEFINED)
-    {
-      if (m_settings.GetLogMissingGenreMappings())
-        Logger::Log(LEVEL_NOTICE, "%s: Could not lookup genre using genre description string instead:'%s'", __FUNCTION__, genreText.c_str());
+    std::string genreText = GetMatchedText(entry.GetPlotOutline(), entry.GetPlot(), genrePattern);
```
Comment (ksooo) [inline src/enigma2/extract/GenreExtractor.cpp]: Effective C++ item 3: Use const whenever possible 

;-)

http://aristeia.com/books.html

File: src/Enigma2.cpp (hunk 5 above)
Comment (ksooo) [reply src/Enigma2.cpp]: You prefer auto over real type (here: std::string) even for non.complex types?

**Comment by phunkyfish** [comment]
Ok, we should be good now

**Comment by phunkyfish** [comment]
Are we good to go?

**Pull request merged**